README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# IndexWizard

<!-- badges: start -->

[![](https://img.shields.io/badge/doi-https://doi.org/10.5281/zenodo.6977265-blue.svg)](https://doi.org/https://doi.org/10.5281/zenodo.6977265)
[![](https://img.shields.io/badge/devel%20version-0.2.1.0-blue.svg)](https://github.com/johannesgeibel/IndexWizard)
[![](https://www.r-pkg.org/badges/version/IndexWizard?color=orange)](https://cran.r-project.org/package=IndexWizard)

[![CRAN
checks](https://badges.cranchecks.info/summary/IndexWizard.svg)](https://cran.r-project.org/web/checks/check_results_IndexWizard.html)
[![](http://cranlogs.r-pkg.org/badges/grand-total/IndexWizard)](https://cran.r-project.org/package=IndexWizard)
[![License:
MIT](https://img.shields.io/badge/license-MIT-blue.svg)](https://cran.r-project.org/web/licenses/MIT)

<!-- badges: end -->

The goal of ‘IndexWizard’ is to provide a framework for the exploration
of effects of complex selection indices on the genetic and phenotypic
progress of a population. It allows to construct selection indices based
on estimated breeding values in animal and plant breeding and to
calculate several analytic measures around. The methodology thereby
allows to analyze genetic gain of traits in the breeding goal that are
not part of the actual index and automatically computes several analytic
measures. It further allows to retrospectively derive realized economic
weights from observed genetic trends. See [Simianer *et al.* “How
economic weights translate into genetic and phenotypic progress, and
vice versa” Genet Sel Evol 55, 38
(2023)](https://doi.org/10.1186/s12711-023-00807-0) for a detailed
description of the methodology and a complex example.

## Installation

You can either install `IndexWizard` from
[CRAN](https://cran.r-project.org/web/packages/IndexWizard/index.html)
the usual way:

``` r
install.packages("IndexWizard")
```

Or directly install the development version of `IndexWizard` from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("johannesgeibel/IndexWizard", build_vignettes = TRUE)
```

Note that `devtools` requires
[Rtools](https://cran.r-project.org/bin/windows/Rtools/) to be installed
on Windows systems.

## Example

All calculations are solely performed in the function `SelInd()`. Its
minimal input requirements are a named genetic covariance matrix (`G`)
for all n breeding goal traits, a named vector of economic weights (`w`)
for the n breeding goal traits and a named vector of reliabilities
(`r2`) of the estimated breeding values for the $m \le n$ index traits.

``` r
library(IndexWizard)
tn <- c("RZM", "RZN", "RZEo") # trait names of breeding goal traits
G <- matrix(
    c(1.0,0.13,0.13,
    0.13,1.0,0.23,
    0.13,0.23,1.0),
    3, 3, dimnames = list(tn,tn)
    ) * 144 # genetic covariance matrix
w <- c(0.7, 0.3, 0) # economic weights, giving weight only to the first two traits
names(w) <- tn
r2 <- c(0.743, 0.673) # reliabilities for the Index traits
names(r2) <- tn[1:2]
SelInd(
  w = w,
  G = G,
  r2 = r2
)
```

Read the vignette `CaseStudy` for a detailed introduction, covering
different use cases of the framework.

``` r
vignette("CaseStudy", package = "IndexWizard")
```

Further find the analysis
[script](https://github.com/johannesgeibel/IndexWizard/blob/main/scripts/makePlots.r)
for [Simianer *et al.*](https://doi.org/10.1186/s12711-023-00807-0) in
this repository.


NEWS.md
# IndexWizard 0.2.1.0

* Added the information about the published paper and made some "beauty changes" to fulfill cran requirements.

# IndexWizard 0.2.0.0

* Switched the calculation of the first derivative of the expected gain to an approximate calculation to account for the constraint that sum(w) has to be 1.


# IndexWizard 0.1.3.2

* Added a `NEWS.md` file to track changes to the package.


R/methods.R
# print -----------------------------------------------------------------------
#' Function to nicely print a SelInd object
#'
#' @param x An object of class SelInd
#' @param ... does nothing, only for compatibility with the generic function
#' 
#' @return No return value, only prints formatted output.
#' 
#' @export
#' @examples
#' tn <- c("RZM", "RZN", "RZEo")
#' G <- matrix(
#'     c(1.0,0.13,0.13,
#'     0.13,1.0,0.23,
#'     0.13,0.23,1.0),
#'     3, 3, dimnames = list(tn,tn)
#'     ) * 144
#' w <- c(0.7, 0.3, 0)
#' names(w) <- tn
#' r2 <- c(0.743, 0.673)
#' names(r2) <- tn[1:2]
#' res <- SelInd(
#'   w = w,
#'   G = G,
#'   r2 = r2
#' )
#' print(res)
#'
print.SelInd <- function(x, ...){
  cat(
    "An object of class SelInd. The index is based on:\n",
    "  - n =",length(x$w),"breeding goal traits\n",
    "  -",sum(x$w!=0),"traits with economic weight != 0\n",
    "  - m =",length(x$r2),"index traits\n")

  # message missing print of matrizes ------------------------------------------
  cat("\n--------------------------------------------------------------------\n")
  mats <- c("G","E","H")
  mats <- mats[!sapply(x[mats],is.null)]
  cat("\nMatrices",paste0(mats,collapse = ", "),"present in SelInd object, but not printed to reduce complexity.\n")
  cat("Extract them by the use of the `$` operator.\n")

  # print variance of index and total excpected gain ---------------------------
  cat("\n--------------------------------------------------------------------\n")
  cat("\nVaiance of the index (`var_I`), expected genetic gain (`dG`)\nand assumed selection intensity (`i`):\n")
  cat("\n$var_I:\n")
  print(round(x$var_I,2))
  if(!is.null(x$dG)){
    cat("\n$dG:\n")
    print(round(x$dG,2))
  }
  if(!is.null(x$i)){
    cat("\n$i:\n")
    print(round(x$i,2))
  }
  # print weights --------------------------------------------------------------
  cat("\n--------------------------------------------------------------------\n")
  cat("\nEconomic (`w`) and index weights (`b`). Potentially they are rescaled (`scaled`)\nso that sum(abs()) = 1. The weights might represent realized (`real`) weights\nbased on an observed composition of the genetic trend:\n")
  cat("\n$w:\n")
  print(round(x$w,2))
  cat("\n$b:\n")
  print(round(x$b,2))
  cat("\n$b_scaled:\n")
  print(round(x$b_scaled,2))
  if(!is.null(x$w_real)){
    cat("\n$w_real:\n")
    print(round(x$w_real,2))
  }
  if(!is.null(x$b_real)){
    cat("\n$b_real:\n")
    print(round(x$b_real,2))
  }


  # print r2 and h2 ------------------------------------------------------------
  cat("\n--------------------------------------------------------------------\n")
  cat("\nreliabilities (`r2`) and heritabilities (`h2`) of the traits:\n")
  cat("\n$r2:\n")
  print(round(x$r2,2))
  if(!is.null(x$h2)){
    cat("\n$h2:\n")
    print(round(x$h2,2))
  }

  # print compositions ---------------------------------------------------------
  cat("\n--------------------------------------------------------------------\n")
  cat("\nComposition (`d`) of genetic (`G`) / phenotypic (`P`) trend.\n")
  cat("The composition might be observed (`obs`) or expected (`exp`).\n")
  cat("The composition might be scaled (`scaled`) so that sum(abs()) = 1:\n")
  if(!is.null(x$d_G_obs)){
    cat("\n$d_G_obs:\n")
    print(round(x$d_G_obs,2))
  }
  if(!is.null(x$d_G_exp)){
    cat("\n$d_G_exp:\n")
    print(round(x$d_G_exp,2))
  }
  if(!is.null(x$d_P_exp)){
    cat("\n$d_P_exp:\n")
    print(round(x$d_P_exp,2))
  }
  if(!is.null(x$d_G_exp_scaled)){
    cat("\n$d_G_exp_scaled:\n")
    print(round(x$d_G_exp_scaled,2))
  }
  if(!is.null(x$d_P_exp_scaled)){
    cat("\n$d_P_exp_scaled:\n")
    print(round(x$d_P_exp_scaled,2))
  }

  # print analytic measures ----------------------------------------------------
  cat("\n--------------------------------------------------------------------\n")
  cat("\nAnalytic measures:\n")
  cat("Correlation between the overall index and the phenotype of trait j (`r_IP`) and\n")
  cat("the loss in prediction accuracy when omitting trait j from the index (`r_IH`).\n")
  cat("Further the approximate first derivative of d_G_exp with respect to w (`del_d_scaled`) with rows\nbeing scaled  so that sum(abs()) = 1.\n")
  if(!is.null(x$r_IP)){
    cat("\n$r_IP:\n")
    print(round(x$r_IP,2))
  }
  if(!is.null(x$r_IH)){
    cat("\n$r_IH:\n")
    print(round(x$r_IH,2))
  }
  if(!is.null(x$del_d_scaled)){
    cat("\n$del_d_scaled:\n")
    print(round(x$del_d_scaled,2))
  }

  #print.default(x)
}

# summary ---------------------------------------------------------------------
#' Function to summarize the content of a SelInd object.
#'
#' @param object An object of class SelInd
#' @param ... does nothing, only for compatibility with the generic function
#'
#' @return No return value, only prints a summary of the SelInd object.
#'
#' @export
#' @examples
#' tn <- c("RZM", "RZN", "RZEo")
#' G <- matrix(
#'     c(1.0,0.13,0.13,
#'     0.13,1.0,0.23,
#'     0.13,0.23,1.0),
#'     3, 3, dimnames = list(tn,tn)
#'     ) * 144
#' w <- c(0.7, 0.3, 0)
#' names(w) <- tn
#' r2 <- c(0.743, 0.673)
#' names(r2) <- tn[1:2]
#' res <- SelInd(
#'   w = w,
#'   G = G,
#'   r2 = r2
#' )
#' summary(res)
#'
summary.SelInd <- function(object, ...){
  cat(
    "An object of class SelInd. The index is based on:\n",
    "  - n =",length(object$w),"breeding goal traits\n",
    "  -",sum(object$w!=0),"traits with economic weight != 0\n",
    "  - m =",length(object$r2),"index traits\n")
  # check whether E is uncorrelated
  check <- any(object$E[upper.tri(object$E)] != 0)
  if(check){
    check <- "correlated"
  }else{
    check <- "uncorrelated"
  }
  cat("\nResidual errors were assumed to be ",check,".\n", sep = "")
  # print non-calculated objects
  obj_in <- paste(names(object)[!sapply(object,is.null)], collapse = ", ")
  obj_out <- paste(names(object)[sapply(object,is.null)], collapse = ", ")
  cat("\nThe SelInd object contains the entries ",obj_in,".\n", sep = "")
  if(any(sapply(object,is.null))){
    obj_out <- paste(names(object)[sapply(object,is.null)], collapse = ", ")
    cat("The SelInd object does not contain the entries ",obj_out,".\n", sep = "")
  }

}

# # as.data.frame ----------------------------------------------------------------
# #' Function to coerce the results of a SelInd object to a data.frame for easy access.
# #'
# #' @param x An object of class SelInd
# #' @param row.names does nothing, only for compatibility with the generic function
# #' @param optional does nothing, only for compatibility with the generic function
# #' @param ... does nothing, only for compatibility with the generic function
# #' @param long logical indicating, whether resulting data.frame shall be in long format.
# #'
# #' @return a data.frame with traits in rows and result vectors in columns. if long = TRUE, a data.frame in long format with three columns (variable, trait and value)
# #' @export
# #'
# as.data.frame.SelInd <- function(x, row.names, optional, ..., long = FALSE){
#   x <- t(x$results)
#   x <- as.data.frame(x)
#   if(long){
#     out <- list()
#     for(i in 1:length(x)){
#       out[[i]] <- data.frame(
#         variable = colnames(x)[i],
#         trait = rownames(x),
#         value = x[[i]]
#       )
#     }
#     out <- do.call(rbind,out)
#     rownames(out) <- NULL
#     return(out)
#   }else{
#     return(x)
#   }
# }



R/SelInd.R
#' Function to calculate selection index
#'
#' @param w Numeric vector of n economic weights. Traits present in G, but not part of the index need to be coded as 0. If traits of G are missing, they will be added automatically with zero weight. Required.
#' @param G Named n*n genetic variance- covariance matrix. Dimnames of G need to match w to ensure correct sorting.  Required.
#' @param r2 Named numeric vector of reliabilities with length m. Required.
#' @param H Named m*m variance-covariance matrix of estimated breeding to internally derive the residual variance-covariance matrix. If H contains more traits than r2, it will be subsetted.
#' @param i Selection intensity
#' @param h2 named numeric vector of length n containing heritabilities for the traits
#' @param d_G_obs named numeric vector of length n containing the observed composition of the genetic gain scaled in genetic standard deviations. If sum(d_G_obs) != 1, it will be rescaled.
#' @param delta small increment to calculate approximate first derivative
#' @param verbose Shall information be printed?
#'
#' @details The framework allows to have less traits in the selection index than in the breeding goal (m < n). Calculation of realized economic weights, however, requires m == n.
#'
#' @return A list of class SelInd
#' @export
#'
#' @examples
#' tn <- c("RZM", "RZN", "RZEo")
#' G <- matrix(
#'     c(1.0,0.13,0.13,
#'     0.13,1.0,0.23,
#'     0.13,0.23,1.0),
#'     3, 3, dimnames = list(tn,tn)
#'     ) * 144
#' w <- c(0.7, 0.3, 0)
#' names(w) <- tn
#' r2 <- c(0.743, 0.673)
#' names(r2) <- tn[1:2]
#' SelInd(
#'   w = w,
#'   G = G,
#'   r2 = r2
#' )
#'
SelInd <- function(
    w,
    G,
    r2,
    H = NULL,
    i = NULL,
    h2 = NULL,
    d_G_obs = NULL,
    delta = 0.0001,
    verbose = TRUE
){
  # initialize central object --------------------------------------------------
  out <- list(
    w = NULL,
    G = NULL,
    E = NULL,
    H = H,
    r2 = r2,
    h2 = h2,
    b = NULL,
    b_scaled = NULL,
    w_real = NULL,
    b_real = NULL,
    var_I = NULL,
    i = i,
    d_G_obs = d_G_obs,
    d_G_exp = NULL,
    d_P_exp = NULL,
    d_G_exp_scaled = NULL,
    d_P_exp_scaled = NULL,
    dG = NULL,
    r_IP = NULL,
    r_IH = NULL,
    del_d_scaled = NULL,
    delta = delta,
    D = NULL # not printed
  )
  class(out) <- "SelInd"

  # check input ----------------------------------------------------------------
  ## w -------------------------------------------------------------------------
  if(!exists("w", inherits = FALSE)){
    stop("No economic weights (w) supplied")
  }else if(is.null(names(w))){
    stop("w is not named")
  }else{
    if(sum(abs(w)) != 1){
      if(verbose) message("- sum(abs(w)) != 1\n  --> rescaling w")
    }
    out$w <- w/sum(abs(w))
  }

  ## G -------------------------------------------------------------------------
  if(!exists("G", inherits = FALSE)){
    stop("No Genetic VCV matrix supplied")
  }else if(dim(G)[1] != dim(G)[2] || any(G[upper.tri(G)] != t(G)[upper.tri(G)])){
    stop("G is not squared")
  }else if(any(colnames(G) != rownames(G))){
    stop("row- and colnames of G do not match")
  }else if(dim(G)[1] < length(w)){
    stop("more traits in w than in G")
  }else if(any(!names(w) %in% colnames(G))){
    stop("Some trait names of w are not in G")
  }else{
    if(any(!colnames(G) %in% names(w))){
      if(verbose) message("- missing traits in w\n  --> adding them with zero weight")
      temp <- rep(0,sum(!colnames(G) %in% names(w)))
      names(temp) <- colnames(G)[!colnames(G) %in% names(w)]
      w <- c(w,temp)
      out$w <- w
    }
    G <- G[names(w),names(w)]
    out$G <- G
  }

  ## r2 and H -------------------------------------------------------------------
  if(is.null(out$r2)){
    stop("No reliabilities supplied")
  }else{
    if(any(!names(out$r2) %in% colnames(out$G))){
      stop("r2 containes traits not in G")
    }
    # set up D
    out$D <- matrix(0,length(out$r2),ncol(out$G),dimnames = list(names(out$r2),colnames(out$G)))
    for(i in 1:length(r2)){
      out$D[names(out$r2)[i],names(out$r2)[i]] <- 1
    }
    # set up R
    R <- diag(out$r2)
    dimnames(R) <- list(names(out$r2), names(out$r2))
    #derive E from r2 and G
    if(is.null(H)){
      g <- diag(out$D %*% out$G %*% t(out$D))
      out$E <- diag((1-r2)/r2*g)
      dimnames(out$E) <- list(names(r2),names(r2))
      if(verbose) message("- only reliabilities given\n  --> setting up E based on r2 and G\n  --> residual errors are assumed to be uncorrelated!")
    }else{
      # include check of H
      if(dim(H)[1] != dim(H)[2] || any(H[upper.tri(H)] != t(H)[upper.tri(H)])){
        stop("H is not squared")
      }else if(any(colnames(H) != rownames(H))){
        stop("row- and colnames of H do not match")
      }else if(dim(H)[1] < length(r2)){
        stop("more traits in r2 than in H")
      }else if(any(!names(r2) %in% colnames(H))){
        stop("Some trait names of r2 are not in H")
      }
      if(verbose) message("- H is given\n  --> setting up E based on r2, G and H\n  --> residual errors are assumed to be correlated!")
      out$H <- H[names(out$r2),names(out$r2)]
      out$E <- solve(sqrt(R)) %*% H %*% solve(sqrt(R))
      out$E <- out$E - out$D %*% out$G %*% t(out$D)
    }
  }


  ## check h2 ------------------------------------------------------------------
  if(!is.null(out$h2)){
    if(length(out$h2) != length(out$w)){
      stop("length of h2 does not equal length of w")
    }else if(any(!names(out$h2) %in% names(out$w))){
      stop("h2 containes traits not in w")
    }else{
      out$h2 <- out$h2[names(out$w)]
    }
  }

  ## check d_G_obs --------------------------------------------------------------
  if(!is.null(out$d_G_obs)){
    if(length(out$d_G_obs) != length(out$r2)){
      stop("length of d_G_obs does not equal length of r2")
    }else if(any(!names(out$d_G_obs) %in% names(out$r2))){
      stop("d_G_obs containes traits not in r2")
    }else{
      out$d_G_obs <- out$d_G_obs[names(out$r2)]
    }
    if(sum(out$d_G_obs) != 1){
      if(verbose) message("- sum(d_G_obs) != 1 -- rescaling d_G_obs")
      out$d_G_obs <- out$d_G_obs/sum(out$d_G_obs)
    }
  }

  # calc index weights ---------------------------------------------------------
  out$b <- solve(R %*% (out$D %*% out$G %*% t(out$D) + out$E) %*% R) %*% R %*% out$D %*% out$G %*% out$w
  out$b_scaled <- out$b / sum(abs(out$b))

  # calc variance of index -----------------------------------------------------
  out$var_I <- t(out$b) %*% R %*% (out$D %*% out$G %*% t(out$D) + out$E) %*% R %*% out$b

  # calc expected composition of genetic and phenotypic trend ---------------------------------
  if(!is.null(out$i)){
    out$d_G_exp <- (out$i / sqrt(out$var_I)[1,1] ) * (out$G %*% t(out$D) %*% R %*% out$b)
    out$dG <- t(out$d_G_exp) %*% out$w
    if(!is.null(out$h2)){
      out$d_P_exp <- out$d_G_exp * sqrt(out$h2) / sqrt(diag(G))
    }else{
      if(verbose) message("- no heritabilities provided\n  --> cannot compute the expected phenotypic trend")
    }
  }else{
    if(verbose) message("- no selection intensity provided\n  --> can only compute the relative genetic and phenotypic trend")
  }
  # relative trend -------------------------------------------------------------
  out$d_G_exp_scaled <- (out$G %*% t(out$D) %*% R %*% out$b)
  out$d_G_exp_scaled <- out$d_G_exp_scaled/sum(abs(out$d_G_exp_scaled))
  if(!is.null(out$h2)){
    # if(!is.null(out$d_G_exp)){
    #   out$d_P_exp <- out$d_G_exp * sqrt(out$h2) / sqrt(diag(G))
    # }
    out$d_P_exp_scaled <- out$d_G_exp_scaled * sqrt(out$h2) / sqrt(diag(G))
    out$d_P_exp_scaled <- out$d_P_exp_scaled/sum(abs(out$d_P_exp_scaled))
  }else{
    if(verbose) message("- no heritabilities provided\n  --> cannot compute the expected relative phenotypic trend")
  }

  # calc analytical measures ---------------------------------------------------
  out$r_IP <- (t(out$b) %*% (R %*% out$D %*% out$G %*% t(out$D))) / (sqrt(out$var_I[1,1] * diag(R %*% out$D %*% out$G %*% t(out$D))))
  out$r_IH <- 1 - sqrt(1 - t(out$b^2) /
                         (t(out$b) %*% R %*% ( out$D %*% out$G %*% t(out$D) + out$E) %*% R %*% out$b %*% diag(solve(R %*% ( out$D %*% out$G %*% t(out$D) + out$E) %*% R))) )

  # first derivative -----------------------------------------------------------
  out$del_d_scaled <- matrix(0,length(out$w),length(out$w),dimnames = list(names(out$w),names(out$w)))
  out$del_d_abs  <- out$del_d_scaled
  d0 <- t(out$G %*% t(out$D) %*% R %*% out$b) / (sqrt(out$var_I)[1,1])

  for(i in 1:length(out$w)){
    # set up modified w
    wmod <- out$w*(1-delta/(1-out$w[i]))
    wmod[i]<-out$w[i]+delta
    
    # get modified b
    bmod <- solve(R %*% (out$D %*% out$G %*% t(out$D) + out$E) %*% R) %*% R %*% out$D %*% out$G %*% wmod
    var_mod <- t(bmod) %*% R %*% (out$D %*% out$G %*% t(out$D) + out$E) %*% R %*% bmod
    
    # calc modified gain
    out$del_d_scaled[i,] <- (out$G %*% t(out$D) %*% R %*% bmod) / sqrt(var_mod)[1,1] # no scaling by i
    out$del_d_abs[i,] <- out$del_d_scaled[i,] - d0 # would be gain different to previous
    out$del_d_scaled[i,] <- out$del_d_abs[i,] /sum(abs(out$del_d_abs[i,]))
    #out$del_d_scaled_new_pook <- out$del_d_scaled_new_diff / diag(out$del_d_scaled_new_diff)
    
  }
  
  # realized weights -----------------------------------------------------

  if(!is.null(out$d_G_obs)){
    if(any(dim(out$G) != dim(out$E))){
      if(verbose) message("- m != n\n  --> subsetting w  and G to length m for calculation of realized weights\n  !!! this discards a part of the variance-covariance structure !!! \n")
      tmp <- colnames(R)
      out$b_real <-  solve(out$G[tmp,tmp] %*% t(out$D[,tmp]) %*% R) %*% out$d_G_obs
      out$w_real <- solve(out$D[,tmp] %*% out$G[tmp,tmp]) %*% (out$D[,tmp] %*% out$G[tmp,tmp] %*% t(out$D[,tmp]) + out$E) %*% solve(out$G[tmp,tmp] %*% t(out$D[,tmp])) %*% out$d_G_obs
    }else{
      out$b_real <-  solve(out$G %*% t(out$D) %*% R) %*% out$d_G_obs
      out$w_real <- solve(out$D %*% out$G) %*% (out$D %*% out$G %*% t(out$D) + out$E) %*% solve(out$G %*% t(out$D)) %*% out$d_G_obs
    }
    out$b_real <- out$b_real/sum(abs(out$b_real))
    out$w_real <- out$w_real/sum(abs(out$w_real))

  }else{
    if(verbose) message("- No observed composition of genetic progress given\n  --> cannot calculate realized weights")
  }

  # return output --------------------------------------------------------------
  out$b <- out$b[,1]
  out$var_I <- out$var_I[1,1]
  if(!is.null(out$b_scaled)) out$b_scaled <- out$b_scaled[,1]
  if(!is.null(out$d_G_exp)) out$d_G_exp <- out$d_G_exp[,1]
  if(!is.null(out$d_P_exp)) out$d_P_exp <- out$d_P_exp[,1]
  if(!is.null(out$d_G_exp_scaled)) out$d_G_exp_scaled <- out$d_G_exp_scaled[,1]
  if(!is.null(out$d_P_exp_scaled)) out$d_P_exp_scaled <- out$d_P_exp_scaled[,1]
  if(!is.null(out$dG)) out$dG <- out$dG[1,1]
  #if(!is.null(out$del_d)) out$del_d <- out$del_d[,1]
  if(!is.null(out$b_real)) out$b_real <- out$b_real[,1]
  if(!is.null(out$w_real)) out$w_real <- out$w_real[,1]
  if(!is.null(out$r_IP)) out$r_IP <- out$r_IP[1,]
  if(!is.null(out$r_IH)) out$r_IH <- out$r_IH[1,]

  return(out)
}


